### PR TITLE
Implement object management

### PR DESCRIPTION
**Pull request by cxdong**
This is the code implementation for the object management. It includes 3 patches, two patches are for the COCONUT-SVSM kernel, which add the Obj trait, ObjHandle and object management. One patch is for the user mode, which add the equivalent ObjHandle define.

The implementation is made according to the object management design code, which has been reviewed at the [PR 4530](https://github.com/coconut-svsm/svsm/pull/453). Thanks to @Freax13 for the invaluable reviews!

### CODE REVIEW

File: kernel/src/syscall/obj.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+impl From<ObjHandle> for u32 {
+    #[inline]
+    fn from(obj_handle: ObjHandle) -> Self {
+        obj_handle.0
+    }
+}
+
+pub type ObjPointer = Arc<dyn Obj>;
```
Comment (Freax13) [inline kernel/src/syscall/obj.rs:54]: I'm not a big fan of the name `ObjPointer`. The word "pointer" reminds me too much of raw pointers and this is not a raw pointer. How about `ObjReference` (or just `Arc<dyn Obj>)`?

**Review by Freax13 — APPROVED** [review]

File: kernel/src/syscall/obj.rs (hunk 1 above)
Comment (cxdong) [reply kernel/src/syscall/obj.rs:54]: I see. Will update the patches and just use Arc<dyn Obj>. Thanks!

File: kernel/src/syscall/obj.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+extern crate alloc;
+
+use crate::cpu::percpu::current_task;
+use crate::error::SvsmError;
+use alloc::sync::Arc;
+
+#[derive(Clone, Copy, Debug)]
+pub enum ObjError {
+    InvalidId,
```
Comment (Freax13) [inline kernel/src/syscall/obj.rs:15]: Let's call this `InvalidHandle` (since we already use `ObjHandle`, not `ObjId`).

**Review by Freax13 — APPROVED** [review]

File: kernel/src/syscall/obj.rs (hunk 2 above)
Comment (cxdong) [reply kernel/src/syscall/obj.rs:15]: I was just writing the comment about this change introduced in this update. Thanks for the review! Will update the name.

Comment (cxdong) [reply kernel/src/syscall/obj.rs]: Just rename to `InvalidHandle`. Now, as this part of code is a bit different with what the document in the [PR453](https://github.com/coconut-svsm/svsm/pull/453) mentioned, I will update the document as well.

File: kernel/src/task/tasks.rs [hunk 3]
```diff
@@ -139,6 +141,9 @@ pub struct Task {
 
     /// Link to scheduler run queue
     runlist_link: LinkedListAtomicLink,
+
+    /// Objects shared among threads within the same process
+    objs: Arc<RWLock<BTreeMap<ObjHandle, Arc<dyn Obj>>>>,
```
Comment (joergroedel) [inline kernel/src/task/tasks.rs:146]: Why is the outer `Arc` needed? Could as well be a `Box`, no?

Comment (Freax13) [reply kernel/src/task/tasks.rs:146]: The `Arc` is needed so that multiple tasks in the same process can share the same object map.

Comment (joergroedel) [reply kernel/src/task/tasks.rs:146]: That makes sense, thanks.

**Pull request merged**